src/file_pipe_log/pipe_builder.rs

```diff
@@ -13,17 +14,69 @@ use crate::event_listener::EventListener;
 use crate::file_builder::FileBuilder;
 use crate::log_batch::LogItemBatch;
 use crate::pipe_log::{FileId, FileSeq, LogQueue};
+use crate::truncate::{TruncateMachine, TruncateQueueParameter};
 use crate::Result;
 
 use super::format::FileNameExt;
 use super::log_file::{build_file_reader, LogFd};
 use super::pipe::{DualPipes, SinglePipe};
 use super::reader::LogItemBatchFileReader;
 
-pub trait ReplayMachine: Send + Default {
-    fn replay(&mut self, item_batch: LogItemBatch, file_id: FileId) -> Result<()>;
+pub trait ReplayMachine: Send + Default + Sync {
+    fn replay(&mut self, item_batch: LogItemBatch, file_id: FileId, path: &Path) -> Result<()>;
 
     fn merge(&mut self, rhs: Self, queue: LogQueue) -> Result<()>;
+
+    fn truncate<B: FileBuilder>(
+        &mut self,
+        _builder: Arc<B>,
+        _truncate_info: &TruncateQueueParameter,
+    ) -> Result<()> {
+        Ok(())
+    }
+}
+
+pub trait ReplayMachineFactory: Sync {
+    type Machine: ReplayMachine;
+
+    fn new_machine(&self) -> Self::Machine {
+        Self::Machine::default()
+    }
+
+    fn truncate<B: FileBuilder>(&self, _: &mut Self::Machine, _: Arc<B>) -> Result<()> {
```

Comment:
This too, you can hold all the necessary information in factory itself.

Reply:
@tabokie After we move the truncate logic to `TruncateMachine`,  the `DefaultMachineFactory` has no function and seems to be useless, It's wise to remove it and revert to original code ?

Reply:
I haven't reviewed your code for the truncate machine. But from the looks of it, you are currently stashing all the log items and truncating them at the end. But initially, I planned to do the truncate **while** reading log items, and doing it one file at a time. For that to happen, you need the truncate parameters during replay, and the customized machine factory.

Reply:
You don't need a new trait method to do the things I described. All parameters and all data can be accessed within the `replay` trait method. You can deduce the file boundary by looking at the `file_id` passed in through `replay` method.
And I don't see how this contradicts with the truncate logic. By "handle file one by one", I mean writing the truncated file immediately after processing the items in it.

Reply:
This the current logic that call replay method 
https://github.com/tikv/raft-engine/blob/517f3b7527412271d6adce969c966ac4567b0e2c/src/file_pipe_log/pipe_builder.rs#L231-L264

1. How can we get the information like queue name, `file_id` can only get the sequence number.  Of course We can put this information in the `TruncateMachine` when use `new_machine` in machine factory.

2. As far as i can see, `relay` method is used to handle one `LogItemBatch`,  As there would be a log of  `LogItemBatch` in one log file, How can we see which one is the last and then do the truncate logic.  The best way i think to do truncate logic in line 142(see above)

3. Handle file one by one, I think, there is still some problem. See the example I mentioned in past, the result will be 
- Front mode
A: 1, 2
B: 7

- Back mode
A:4, 5
B: 10

- All mode
A: (empty)
B: (empty)

Index of raft groups will still discontinuous in mode front and back. is it what we expected?

Reply:
1. Yes, Putting truncate information into truncate machine was the initial plan.
2. "You can deduce the file boundary by looking at the file_id passed in through replay method." For sequential replay machine, you keep a record of last replayed file id, and when incoming file id == last replayed id + 1, you know it belongs to a new file.
3. There is no problem for back and all mode. For back mode you can maintain a record of last contiguous index, and delete all entries larger than it. But I do see a problem with front mode that I didn't notice before. You can't deduce the latest logical hole by scanning forward. How about we add a bool parameter `backward` to the `scan` method, so that we can scan the files backward?

Reply:
@tabokie 
> There is no problem for back and all mode

You mean result of mode back and all in https://github.com/tikv/raft-engine/pull/175#discussion_r779275658 is reasonable?, If so the only things left is implementation.

> But I do see a problem with front mode that I didn't notice before. You can't deduce the latest logical hole by scanning forward. How about we add a bool parameter backward to the scan method, so that we can scan the files backward?

Indeed, we need a method to traverse file reversely, let me have a try

Reply:
No, you can accomplish the old results (https://github.com/tikv/raft-engine/pull/175#discussion_r778799444) with new design. "For back mode you can maintain a record of last contiguous index, and delete all entries larger than it."

Reply:
@tabokie, Here is the situation
> 1. "You can deduce the file boundary by looking at the file_id passed in through replay method." For sequential replay machine, you keep a record of last replayed file id, and when incoming file id == last replayed id + 1, you know it belongs to a new file.

We may need another trait the flush the data of last file, As the content can only be wrote to file when handle the next file, we need a trait to save the last file. Is it element enough? or do you have any suggestions

2.  For `Back` mode we need a reverse iterator just as the `Front` mode, But for `All` mode, I'am afraid we should keep all data(items) in memory temporarily and check if there is a hole, Then we can safely save the items to log files. Is there any wrong with the point?

Looking forward your reply, thanks